File: jsfring-webservice/src/main/java/fr/pinguet62/jsfring/webservice/converter/UserMapper.java
package fr.pinguet62.jsfring.webservice.converter;

import com.querydsl.core.types.Predicate;
import fr.pinguet62.jsfring.model.sql.Profile;
import fr.pinguet62.jsfring.model.sql.QProfile;
import fr.pinguet62.jsfring.model.sql.User;
import fr.pinguet62.jsfring.service.ProfileService;
import fr.pinguet62.jsfring.webservice.config.MapstructConfig;
import fr.pinguet62.jsfring.webservice.dto.UserDto;
import org.mapstruct.Mapper;
import org.mapstruct.MappingTarget;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.List;
import java.util.Set;

import static java.util.stream.Collectors.toSet;

@Mapper(config = MapstructConfig.class)
public abstract class UserMapper {

    @Autowired
    private ProfileService profileService;

    protected Set<Profile> keyToEntity(Set<Integer> keys) {
        Predicate predicate = QProfile.profile.id.in(keys);
        return profileService.findAll(predicate).collect(toSet()).block();
    }

    protected Set<Integer> entityToKey(Set<Profile> keys) {
        return keys.stream().map(Profile::getId).collect(toSet());
    }

    public abstract User fromDto(UserDto dto);

    public abstract UserDto toDto(User entity);

    public abstract List<UserDto> toDto(List<User> entities);

    public abstract void updateFromDto(User entity, @MappingTarget UserDto dto);

}

File: jsfring-webservice/src/main/kotlin/fr/pinguet62/jsfring/webservice/converter/UserMapper.kt
package fr.pinguet62.jsfring.webservice.converter

import com.querydsl.core.types.Predicate
import fr.pinguet62.jsfring.model.sql.Profile
import fr.pinguet62.jsfring.model.sql.QProfile
import fr.pinguet62.jsfring.model.sql.User
import fr.pinguet62.jsfring.service.ProfileService
import fr.pinguet62.jsfring.webservice.config.MapstructConfig
import fr.pinguet62.jsfring.webservice.dto.UserDto
import org.mapstruct.Mapper
import org.mapstruct.MappingTarget
import org.springframework.beans.factory.annotation.Autowired
import java.util.stream.Collectors.toSet

@Mapper(config = MapstructConfig::class)
abstract class UserMapper {

    @Autowired
    lateinit var profileService: ProfileService

    protected fun keyToEntity(keys: Set<Int>): Set<Profile> {
        var predicate: Predicate = QProfile.profile.id.`in`(keys)
        return profileService.findAll(predicate).collect(toSet()).block()
    }

    protected fun entityToKey(keys: Set<Profile>): Set<Int> =
            keys.map { it.id }.toSet()

    abstract fun fromDto(dto: UserDto): User

    abstract fun toDto(entity: User): UserDto

    abstract fun toDto(entities: List<User>): List<UserDto>

    abstract fun updateFromDto(entity: User, @MappingTarget dto: UserDto)

}
